# Sprint 3 – Data Cleaning & Data Preparation

## Objetivo

Limpiar, validar y preparar el dataset de RetailPulse para garantizar que los datos utilizados en los análisis posteriores sean consistentes, reproducibles y adecuados para SQL y Power BI.

In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns",None)


In [3]:
df = pd.read_csv( "../data/raw/Sample - Superstore.csv",
    encoding="latin1"
)
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [4]:
df.shape

(9994, 21)

Paso 3. Crear una copia de trabajo



In [5]:
df_clean = df.copy()

In [6]:
df_clean.dtypes

Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code        int64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object

In [7]:
df_clean["Order Date"] = pd.to_datetime(df_clean["Order Date"])
df_clean["Ship Date"] = pd.to_datetime(df_clean["Ship Date" ] )

In [8]:
df_clean.dtypes

Row ID                    int64
Order ID                    str
Order Date       datetime64[us]
Ship Date        datetime64[us]
Ship Mode                   str
Customer ID                 str
Customer Name               str
Segment                     str
Country                     str
City                        str
State                       str
Postal Code               int64
Region                      str
Product ID                  str
Category                    str
Sub-Category                str
Product Name                str
Sales                   float64
Quantity                  int64
Discount                float64
Profit                  float64
dtype: object

Las columnas Order Date y Ship Date fueron convertidas del tipo string a datetime para facilitar el análisis temporal del negocio. Esta transformación permitirá calcular indicadores por año, trimestre, mes o día, además de analizar tiempos de envío y tendencias de ventas a lo largo del tiempo.

In [9]:
df_clean.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   str           
 2   Order Date     9994 non-null   datetime64[us]
 3   Ship Date      9994 non-null   datetime64[us]
 4   Ship Mode      9994 non-null   str           
 5   Customer ID    9994 non-null   str           
 6   Customer Name  9994 non-null   str           
 7   Segment        9994 non-null   str           
 8   Country        9994 non-null   str           
 9   City           9994 non-null   str           
 10  State          9994 non-null   str           
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   str           
 13  Product ID     9994 non-null   str           
 14  Category       9994 non-null   str           
 15  Sub-Category   9994 non-null   s

In [10]:
categorial_colums = ["Ship Mode", "Segment","Country","City","State","Region","Category","Sub-Category"]

for col in categorial_colums:
    print(f"{col}:{df_clean[col].nunique()} valores unicos")

Ship Mode:4 valores unicos
Segment:3 valores unicos
Country:1 valores unicos
City:531 valores unicos
State:49 valores unicos
Region:4 valores unicos
Category:3 valores unicos
Sub-Category:17 valores unicos


## Identificación de variables categóricas

Antes de convertir las columnas al tipo `category`, se evaluó el número de valores únicos de cada variable.

Las columnas seleccionadas presentan un número reducido de categorías repetidas, por lo que su conversión permitirá optimizar el consumo de memoria sin alterar la información almacenada.

Las columnas `Customer Name`, `Product Name`, `Customer ID` y `Product ID` se mantendrán como texto debido a su alta cardinalidad o porque funcionan como identificadores.

In [11]:
categorial_colums = ["Ship Mode", "Segment","Country","City","State","Region","Category","Sub-Category"]
for col in categorial_colums:
    df_clean[col] = df_clean[col].astype("category")


In [12]:
df_clean.dtypes

Row ID                    int64
Order ID                    str
Order Date       datetime64[us]
Ship Date        datetime64[us]
Ship Mode              category
Customer ID                 str
Customer Name               str
Segment                category
Country                category
City                   category
State                  category
Postal Code               int64
Region                 category
Product ID                  str
Category               category
Sub-Category           category
Product Name                str
Sales                   float64
Quantity                  int64
Discount                float64
Profit                  float64
dtype: object

In [13]:
df_clean.info(memory_usage="deep")

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9994 non-null   int64         
 1   Order ID       9994 non-null   str           
 2   Order Date     9994 non-null   datetime64[us]
 3   Ship Date      9994 non-null   datetime64[us]
 4   Ship Mode      9994 non-null   category      
 5   Customer ID    9994 non-null   str           
 6   Customer Name  9994 non-null   str           
 7   Segment        9994 non-null   category      
 8   Country        9994 non-null   category      
 9   City           9994 non-null   category      
 10  State          9994 non-null   category      
 11  Postal Code    9994 non-null   int64         
 12  Region         9994 non-null   category      
 13  Product ID     9994 non-null   str           
 14  Category       9994 non-null   category      
 15  Sub-Category   9994 non-null   c

## Optimización del uso de memoria

Se evaluó la cardinalidad de las variables categóricas y se identificaron ocho columnas susceptibles de convertirse al tipo `category`.

Las columnas convertidas fueron:

- Ship Mode
- Segment
- Country
- City
- State
- Region
- Category
- Sub-Category

Antes de la optimización, el conjunto de datos consumía aproximadamente **9.2 MB** de memoria.

Después de convertir estas variables al tipo `category`, el consumo se redujo a **4.3 MB**, lo que representa una disminución aproximada del **53 %**.

Esta optimización mejora el rendimiento del procesamiento de datos sin modificar la información contenida en el dataset.

In [14]:
(df_clean["Sales"]<=0).sum()

np.int64(0)

### Regla de negocio 1: Ventas positivas

Se verificó que todas las ventas fueran mayores que cero.

La validación no identificó registros con ventas negativas o iguales a cero, lo que indica que las transacciones cumplen con la regla de negocio esperada y no requieren acciones correctivas en esta etapa.

Resultado: 0

In [15]:


df_clean[
    (df_clean["Discount"] < 0) |
    (df_clean["Discount"] > 1)
]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit


### Regla de negocio 2: Validación de descuentos

Se verificó que todos los valores de la variable `Discount` estuvieran dentro del rango permitido (0 a 1).

La validación no identificó registros con descuentos negativos o superiores al 100 %, por lo que se concluye que la variable cumple con las reglas de negocio esperadas y no requiere correcciones.

In [19]:
(df_clean["Quantity"]<=0).sum()

np.int64(0)

### Regla de negocio 3: Validación de cantidad vendida

Se verificó que todas las transacciones registraran una cantidad de productos mayor que cero.

No se identificaron registros con cantidades iguales o inferiores a cero, por lo que la variable `Quantity` cumple con las reglas de negocio esperadas y puede utilizarse con confianza en los análisis posteriores.

In [20]:
df_clean[df_clean["Ship Date"] < df_clean["Order Date"]]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit


### Regla de negocio 4: Coherencia de fechas

Se verificó que la fecha de envío (`Ship Date`) no fuera anterior a la fecha de realización del pedido (`Order Date`).

No se identificaron registros que incumplieran esta condición. Por lo tanto, las fechas presentan una secuencia temporal consistente con el proceso operativo del negocio.

In [21]:
(df_clean["Profit"]< 0).sum()

np.int64(1871)

### Análisis de transacciones con pérdidas

Se identificaron **1,871 líneas de venta con utilidad negativa**, equivalentes aproximadamente al **18.7 % de los 9,994 registros del dataset**.

A diferencia de otras validaciones realizadas, una utilidad negativa no se considera automáticamente un error de calidad de datos. Estos registros pueden representar operaciones comerciales reales en las que los costos asociados superaron los ingresos generados.

Por esta razón, los registros con `Profit < 0` se conservarán y serán analizados posteriormente para determinar si las pérdidas están relacionadas con factores como descuentos, categorías, subcategorías, productos o regiones.

In [22]:
df_clean.nsmallest(10,"Profit")[  
[
    "Order ID",
        "Product Name",
        "Category",
        "Sub-Category",
        "Sales",
        "Quantity",
        "Discount",
        "Profit"
 ]
]

,Order ID,Product Name,Category,Sub-Category,Sales,Quantity,Discount,Profit
7772,CA-2016-108196,Cubify CubeX 3D Printer Double Head Print,Technology,Machines,4499.985,5,0.7,-6599.9780
683,US-2017-168116,Cubify CubeX 3D Printer Triple Head Print,Technology,Machines,7999.980,4,0.5,-3839.9904
9774,CA-2014-169019,GBC DocuBind P400 Electric Binding System,Office Supplies,Binders,2177.584,8,0.8,-3701.8928
3011,CA-2017-134845,Lexmark MX611dhe Monochrome Laser Printer,Technology,Machines,2549.985,5,0.7,-3399.9800
4991,US-2017-122714,Ibico EPK-21 Electric Binding System,Office Supplies,Binders,1889.990,5,0.8,-2929.4845
3151,CA-2015-147830,Cubify CubeX 3D Printer Double Head Print,Technology,Machines,1799.994,2,0.7,-2639.9912
5310,CA-2017-131254,Fellowes PB500 Electric Punch Plastic Comb Bin...,Office Supplies,Binders,1525.188,6,0.8,-2287.7820
9639,CA-2015-116638,Chromcraft Bull-Nose Wood Oval Conference Tabl...,Furniture,Tables,4297.644,13,0.4,-1862.3124
1199,CA-2016-130946,GBC DocuBind P400 Electric Binding System,Office Supplies,Binders,1088.792,4,0.8,-1850.9464
2697,CA-2014-145317,Cisco TelePresence System EX90 Videoconferenci...,Technology,Machines,22638.480,6,0.5,-1811.0784


### Evaluación de valores extremos en Profit

Se analizaron las 10 líneas de venta con mayores pérdidas para determinar si los valores extremos identificados en `Profit` correspondían a posibles errores de calidad.

Los registros muestran transacciones comercialmente plausibles y se observa que varias de las mayores pérdidas están asociadas con descuentos elevados, principalmente en las subcategorías `Machines`, `Binders` y `Tables`.

Por ejemplo, la transacción con mayor pérdida presenta ventas aproximadas de $4,500, un descuento del 70 % y una utilidad cercana a -$6,600.

Por lo tanto, los valores negativos extremos de `Profit` no serán eliminados durante la limpieza. Se conservarán para el análisis exploratorio posterior, ya que podrían representar información relevante sobre problemas de rentabilidad.

A partir de esta revisión se plantea la hipótesis de que los descuentos elevados podrían estar asociados con una reducción significativa de la rentabilidad. Esta relación será evaluada posteriormente mediante el análisis del conjunto completo de datos.

In [23]:
df_clean.duplicated().sum()

np.int64(0)

### Validación de registros duplicados

Se verificó la existencia de registros completamente duplicados mediante `duplicated()`.

El análisis identificó **0 registros duplicados**, por lo que no fue necesario eliminar observaciones.

Es importante distinguir entre registros duplicados y valores repetidos en `Order ID`, ya que una misma orden puede contener múltiples productos y, por lo tanto, aparecer en varias filas del dataset.

In [24]:
df_clean.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

### Validación de valores nulos

Se verificó nuevamente la presencia de valores nulos después de las transformaciones realizadas durante el proceso de limpieza.

Las 21 variables presentan **0 valores nulos**, por lo que no fue necesario aplicar técnicas de eliminación o imputación de datos faltantes.

Además, las transformaciones realizadas durante la preparación de los datos no generaron nuevos valores nulos.

In [25]:
df_clean["Row ID"].nunique()

9994

### Validación del identificador Row ID

Se verificó la unicidad de la variable `Row ID`.

El dataset contiene 9,994 registros y se identificaron 9,994 valores únicos de `Row ID`, confirmando que cada fila cuenta con un identificador único.

No se requieren modificaciones sobre esta variable.

In [26]:
df_clean["Order ID"].nunique()

5009

### Validación de Order ID

El dataset contiene 9,994 líneas de venta correspondientes a **5,009 pedidos únicos**.

La diferencia entre el número de registros y el número de pedidos confirma que una misma orden puede contener múltiples productos. Por esta razón, la repetición de `Order ID` es consistente con la estructura del negocio y no representa un problema de duplicidad.

Para futuros indicadores relacionados con el número de pedidos se utilizará el conteo de valores únicos de `Order ID`, evitando utilizar directamente el número de filas del dataset.

In [29]:
df_clean["Customer ID"].nunique()


793

In [30]:
df_clean["Product ID"].nunique()

1862

### Validación de identificadores y granularidad

Se analizaron los principales identificadores del dataset para comprender la granularidad de la información.

Se identificaron:

- 9,994 líneas de venta.
- 5,009 pedidos únicos.
- 793 clientes únicos.
- 1,862 productos identificados mediante `Product ID`.

Los resultados confirman que el dataset se encuentra a nivel de línea de pedido. Un cliente puede realizar múltiples pedidos y cada pedido puede contener múltiples productos.

Esta estructura deberá considerarse durante la construcción de indicadores para evitar utilizar el número de filas como equivalente al número de pedidos o clientes.

In [32]:
df_clean["Product Name"].nunique()

1850

In [33]:
df_clean.groupby("Product Name",observed=True)["Product ID"].nunique().sort_values(ascending=False).head(10)

Product Name
Staples                                    10
Staple envelope                             9
Easy-staple paper                           8
Staples in misc. colors                     7
Staple holder                               3
Staple remover                              3
#10- 4 1/8" x 9 1/2" Recycled Envelopes     2
Avery Non-Stick Binders                     2
Eldon Wave Desk Accessories                 2
KI Adjustable-Height Table                  2
Name: Product ID, dtype: int64

### Validación de identificadores de producto

Se identificaron **1,862 valores únicos de `Product ID`** y **1,850 nombres de producto únicos**.

Al analizar la relación entre ambas variables, se observó que algunos nombres de producto están asociados con múltiples identificadores. Por ejemplo, `Staples` aparece relacionado con 10 `Product ID` diferentes y `Staple envelope` con 9.

Esta situación no se considera automáticamente un error de calidad, ya que podría representar diferentes SKU, presentaciones o variantes comerciales agrupadas bajo un mismo nombre.

Debido a que el dataset no incluye información adicional del catálogo maestro que permita validar esta relación, se decidió conservar los identificadores originales y documentar esta característica para futuros análisis.

Para análisis a nivel de producto se deberá especificar si la unidad de análisis corresponde a `Product ID` o `Product Name`.

In [ ]:
df_clean["Shipping Days"]  = (
    df_clean["Ship Date"]- df_clean["Order Date"]
).dt.days


0    3
1    3
2    4
3    7
4    7
Name: Shipping Days, dtype: int64

In [39]:
df_clean[
    ["Order Date", "Ship Date", "Shipping Days"]
].head(10)

,Order Date,Ship Date,Shipping Days
0,2016-11-08,2016-11-11,3
1,2016-11-08,2016-11-11,3
2,2016-06-12,2016-06-16,4
3,2015-10-11,2015-10-18,7
4,2015-10-11,2015-10-18,7
5,2014-06-09,2014-06-14,5
6,2014-06-09,2014-06-14,5
7,2014-06-09,2014-06-14,5
8,2014-06-09,2014-06-14,5
9,2014-06-09,2014-06-14,5


In [40]:
df_clean["Shipping Days"].describe()

count    9994.000000
mean        3.958175
std         1.747567
min         0.000000
25%         3.000000
50%         4.000000
75%         5.000000
max         7.000000
Name: Shipping Days, dtype: float64

### Creación de la variable Shipping Days

Se creó la variable `Shipping Days` calculando la diferencia entre `Ship Date` y `Order Date`.

El análisis descriptivo mostró un tiempo promedio de aproximadamente **3.96 días**, con una mediana de **4 días**. El tiempo mínimo registrado fue de 0 días y el máximo de 7 días.

La variable permitirá analizar posteriormente el desempeño operativo de los envíos y comparar los tiempos de procesamiento según modalidad de envío, región u otras dimensiones del negocio.

Debido a que el dataset se encuentra a nivel de línea de pedido, los indicadores definitivos de tiempo de envío a nivel de orden deberán calcularse utilizando pedidos únicos.

In [41]:
(df_clean["Shipping Days"]<0).sum()

np.int64(0)

In [45]:
df_clean["Shipping Days"].value_counts().sort_index()

Shipping Days
0     519
1     369
2    1334
3    1005
4    2774
5    2169
6    1203
7     621
Name: count, dtype: int64

#### Validación de Shipping Days

La variable `Shipping Days` presenta valores comprendidos entre 0 y 7 días, sin identificarse tiempos negativos.

El tiempo de envío más frecuente es de **4 días**, presente en 2,774 líneas de venta. La distribución se concentra principalmente entre 3 y 5 días.

Los registros con tiempos de envío más altos se conservarán para analizar posteriormente si existen diferencias asociadas con la modalidad de envío, región u otras características operativas.

In [46]:
df_clean["Order Year"]= df_clean["Order Date"].dt.year

In [49]:
df_clean["Order Year"].value_counts().sort_index()

Order Year
2014    1993
2015    2102
2016    2587
2017    3312
Name: count, dtype: int64

### Creación de la variable Order Year

Se creó la variable `Order Year` a partir de `Order Date` para facilitar el análisis temporal.

El dataset contiene información correspondiente al periodo **2014–2017**. El número de líneas de venta registradas aumenta progresivamente a lo largo del periodo, pasando de 1,993 registros en 2014 a 3,312 en 2017.

Este resultado describe el volumen de registros y no debe interpretarse directamente como crecimiento de ventas o pedidos. Dichos indicadores serán evaluados posteriormente utilizando el valor de `Sales` y el número de `Order ID` únicos.

In [66]:
df_clean["Order Month"] = df_clean["Order Date"].dt.month

df_clean["Order_Month_Name"] = df_clean["Order Date"].dt.month_name()

In [67]:
df_clean[
   ["Order Date","Order Year","Order_Month_Name","Order Month"]


].head(10)

,Order Date,Order Year,Order_Month_Name,Order Month
0,2016-11-08,2016,November,11
1,2016-11-08,2016,November,11
2,2016-06-12,2016,June,6
3,2015-10-11,2015,October,10
4,2015-10-11,2015,October,10
5,2014-06-09,2014,June,6
6,2014-06-09,2014,June,6
7,2014-06-09,2014,June,6
8,2014-06-09,2014,June,6
9,2014-06-09,2014,June,6


In [68]:
df_clean["Order Month"].value_counts().sort_index()

Order Month
1      381
2      300
3      696
4      668
5      735
6      717
7      710
8      706
9     1383
10     819
11    1471
12    1408
Name: count, dtype: int64

### Creación de variables mensuales

Se crearon las variables `Order Month` y `Order Month Name` a partir de `Order Date`.

`Order Month` almacena el número del mes y permitirá mantener un orden cronológico correcto, mientras que `Order Month Name` proporciona una etiqueta legible para futuras visualizaciones.

Se verificó la distribución de registros entre los 12 meses. No se identificaron meses fuera del rango esperado.

Aunque se observa una mayor cantidad de líneas de venta en determinados meses, este resultado no se interpreta todavía como mayor volumen de ventas, ya que será necesario analizar posteriormente el valor de `Sales`.

In [69]:
df_clean["Order Quarter"] = df_clean["Order Date"].dt.quarter

In [70]:
df_clean["Order Quarter"].value_counts().sort_index()

Order Quarter
1    1377
2    2120
3    2799
4    3698
Name: count, dtype: int64

### Creación de Order Quarter

Se creó la variable `Order Quarter` a partir de `Order Date` para clasificar cada registro según el trimestre en que se realizó la operación.

La validación confirmó que todos los registros fueron clasificados dentro de los cuatro trimestres esperados.

Se observa preliminarmente una mayor concentración de líneas de venta durante el cuarto trimestre. Sin embargo, esta distribución no se interpreta directamente como mayor volumen de ventas o rentabilidad, ya que dichos indicadores serán evaluados posteriormente utilizando `Sales`, `Profit` y pedidos únicos.

In [71]:
df_clean.shape

(9994, 27)

In [72]:
df_clean.columns.tolist()

['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit',
 'Shipping Days',
 'Order Year',
 'Order Month',
 'Order Month Name',
 'Order_Month_Name',
 'Order Quarter']

In [73]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 27 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Row ID            9994 non-null   int64         
 1   Order ID          9994 non-null   str           
 2   Order Date        9994 non-null   datetime64[us]
 3   Ship Date         9994 non-null   datetime64[us]
 4   Ship Mode         9994 non-null   category      
 5   Customer ID       9994 non-null   str           
 6   Customer Name     9994 non-null   str           
 7   Segment           9994 non-null   category      
 8   Country           9994 non-null   category      
 9   City              9994 non-null   category      
 10  State             9994 non-null   category      
 11  Postal Code       9994 non-null   int64         
 12  Region            9994 non-null   category      
 13  Product ID        9994 non-null   str           
 14  Category          9994 non-null   c

In [75]:
df_clean.drop(
    columns=["Order_Month_Name"],
    inplace=True
)

In [76]:
df_clean.columns.tolist()

['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit',
 'Shipping Days',
 'Order Year',
 'Order Month',
 'Order Month Name',
 'Order Quarter']

In [77]:
final_validation = {
    "Filas": len(df_clean),
    "Columnas": df_clean.shape[1],
    "Valores nulos": df_clean.isnull().sum().sum(),
    "Filas duplicadas": df_clean.duplicated().sum(),
    "Ventas <= 0": (df_clean["Sales"] <= 0).sum(),
    "Cantidades <= 0": (df_clean["Quantity"] <= 0).sum(),
    "Descuentos inválidos": (
        (df_clean["Discount"] < 0) |
        (df_clean["Discount"] > 1)
    ).sum(),
    "Fechas inconsistentes": (
        df_clean["Ship Date"] < df_clean["Order Date"]
    ).sum(),
    "Shipping Days negativos": (
        df_clean["Shipping Days"] < 0
    ).sum()
}

pd.Series(final_validation)

Filas                      9994
Columnas                     26
Valores nulos                 0
Filas duplicadas              0
Ventas <= 0                   0
Cantidades <= 0               0
Descuentos inválidos          0
Fechas inconsistentes         0
Shipping Days negativos       0
dtype: int64

## Validación final de calidad

Después de realizar las transformaciones y validaciones del Sprint 3, se ejecutó un control final sobre el dataset preparado.

Resultados:

- Registros: 9,994
- Variables: 26
- Valores nulos: 0
- Filas completamente duplicadas: 0
- Ventas iguales o menores a cero: 0
- Cantidades iguales o menores a cero: 0
- Descuentos fuera del rango 0–1: 0
- Fechas de envío anteriores a la fecha del pedido: 0
- Tiempos de envío negativos: 0

No se identificaron registros que requirieran eliminación bajo las reglas de calidad establecidas.

Los registros con utilidad negativa fueron conservados debido a que representan posibles operaciones comerciales reales y serán investigados durante el análisis exploratorio.

## Exportación del dataset limpio

Después de completar las transformaciones y validaciones de calidad, se exportará el dataset procesado para utilizarlo en las siguientes etapas del proyecto.

In [79]:
df_clean.to_csv("../data/processed/superstore_clean.csv",index=False)

In [80]:
df_test = pd.read_csv(
    "../data/processed/superstore_clean.csv"
)

df_test.shape

(9994, 26)

In [81]:
df_test.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Shipping Days,Order Year,Order Month,Order Month Name,Order Quarter
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,3,2016,11,November,4
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,3,2016,11,November,4
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,4,2016,6,June,2
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,7,2015,10,October,4
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,7,2015,10,October,4


In [83]:
df_test.isnull().sum().sum()

np.int64(0)

In [84]:
df_test.columns.tolist()

['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit',
 'Shipping Days',
 'Order Year',
 'Order Month',
 'Order Month Name',
 'Order Quarter']

In [85]:
df_test.dtypes

Row ID                int64
Order ID                str
Order Date              str
Ship Date               str
Ship Mode               str
Customer ID             str
Customer Name           str
Segment                 str
Country                 str
City                    str
State                   str
Postal Code           int64
Region                  str
Product ID              str
Category                str
Sub-Category            str
Product Name            str
Sales               float64
Quantity              int64
Discount            float64
Profit              float64
Shipping Days         int64
Order Year            int64
Order Month           int64
Order Month Name        str
Order Quarter         int64
dtype: object